streams/src/main/java/org/apache/kafka/streams/processor/TaskId.java

```diff
@@ -90,7 +220,9 @@ public boolean equals(Object o) {
 
         if (o instanceof TaskId) {
             TaskId other = (TaskId) o;
-            return other.topicGroupId == this.topicGroupId && other.partition == this.partition;
+            return other.topicGroupId == this.topicGroupId && other.partition == this.partition && 
+                   other.numberOfInputPartitions == this.numberOfInputPartitions &&
+                   other.numberOfStateStores == this.numberOfStateStores && mapsEqual(other);
```

Comment:
Equals doesn't match hashcode, which is a bug. Also, it's generally a bug for mutable fields to be used in equals or hashcode.

Reply:
I agree that both need to the same. Note, that even if `numberOfStateStores` is not `final`, it is actually not mutable (ie, we know it will never change) -- it just not `final` to simplify the deserialization code. If you want them to be `final`, that would be possible I think.

However, thinking about this twice, I am not sure if the existing code would be sufficient? In fact, if would be a bug if there are two `TaskId` instances with same `topicGroupId` and same `partition` (what is the task ID) but different other fields...

Reply:
I mostly agree with what you mentioned. From what was discussed, it appears that the best option is like what @vvcephei said: to move the metadata stored originally in ```TaskId``` into another class. Whether it is stored in SubscriptionInfo or some other class is still undecided. WDYT @mjsax  @vvcephei ? Storing the information in TaskId is relatively simple, especially since when an ```AbstractTask``` is initialized, the necessary data is readily available. However, storing it in other classes might require some more work.

Reply:
I didn't realize that `TaskId` is public -- we should not change the interface to avoid a KIP. I like the proposal to add `TaskMetadata` class.